keras_nlp/layers/preprocessing/mlm_mask_generator.py

```diff
@@ -0,0 +1,223 @@
+        array([[1, 0, 0, 0, 0],
+               [1, 1, 0, 0, 0]])>}
+    """
+
+    def __init__(
+        self,
+        vocabulary_size,
+        mask_selection_rate,
+        mask_selection_length=None,
+        unselectable_token_ids=None,
+        mask_token_id=0,
+        mask_token_rate=0.8,
+        random_token_rate=0.1,
+        padding_token_id=None,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.vocabulary_size = vocabulary_size
+        self.unselectable_token_ids = unselectable_token_ids
+        self.mask_selection_rate = mask_selection_rate
+        self.mask_selection_length = mask_selection_length
+        self.mask_token_rate = mask_token_rate
+        self.random_token_rate = random_token_rate
+        self.padding_token_id = padding_token_id
+
+        if mask_token_id >= vocabulary_size:
+            raise ValueError(
+                f"Mask token id should be in range [0, vocabulary_size - 1], "
+                f"but received mask_token_id={mask_token_id}."
+            )
+        self.mask_token_id = mask_token_id
+
+        max_selections = self.mask_selection_length
+        if max_selections is None:
+            # Set a large number to remove the `max_selections_per_batch` cap.
+            max_selections = 2**31 - 1
+        self._random_selector = tf_text.RandomItemSelector(
+            max_selections_per_batch=max_selections,
+            selection_rate=self.mask_selection_rate,
+            unselectable_ids=self.unselectable_token_ids,
+        )
+        self._mask_values_chooser = tf_text.MaskValuesChooser(
+            self.vocabulary_size,
+            self.mask_token_id,
+            mask_token_rate=self.mask_token_rate,
+            random_token_rate=self.random_token_rate,
+        )
+
+    def call(self, inputs):
+        input_is_ragged = isinstance(inputs, tf.RaggedTensor)
+        input_is_1d = tf.rank(inputs) == 1
+        if input_is_1d:
+            # If inputs is of rank 1, we manually add the batch axis.
+            inputs = inputs[tf.newaxis, :]
+        if not input_is_ragged:
+            # Convert to RaggedTensor to avoid masking out padded token.
+            inputs = tf.RaggedTensor.from_tensor(
+                inputs,
+                padding=self.padding_token_id,
+            )
+        (tokens, masked_positions, masked_ids,) = tf_text.mask_language_model(
+            inputs,
+            item_selector=self._random_selector,
+            mask_values_chooser=self._mask_values_chooser,
+        )
+
+        if not input_is_ragged:
+            # If inputs is a Tensor not RaggedTensor, we format the masked
+            # output to be a Tensor.
+            tokens = tokens.to_tensor()
+
+        if self.mask_selection_length:
+            # Pad `masked_positions`, `mask_weights` and `masked_ids` if there
+            # is a desired `mask_selection_length` set.
+            masked_positions, mask_weights = tf_text.pad_model_inputs(
```

Comment:
I think here as well we should make sure the mask_weights is of compute_dtype. It might be easier and simpler to just call `x.to_tensor(shape)` here directly. It's very simple code, and we don't actually need the `merge_dims` call that tf_text is doing.

https://github.com/tensorflow/text/blob/v2.8.1/tensorflow_text/python/ops/pad_model_inputs_ops.py#L79

If we write it ourselves, we could just do

```
mask_weights = array_ops.ones_like(masked_positions, dtype=self.compute_dtype)
mask_weights = mask_weights.to_tensor(shape=target_shape)
```

Reply:
array_ops and math_ops are not exposed as public python api, shall we avoid using direct imports from tensorflow? I am now using  a `tf.cast()` to cast the `mask_weights`.

Reply:
I think overall, a cleaner replacement for this entire block looks like this I think...

```python
mask_weights = tf.ones_like(mask_positions, self.compute_dtype)
# If mask_selection_length, covert to raggeds to dense.
if self.mask_selection_length:
    target_shape = tf.cast([-1, self.mask_selection_length], dtypes.int64)
    mask_positions = mask_positions.to_tensor(shape=target_shape)
    mask_ids = mask_ids.to_tensor(shape=target_shape)
    mask_weights = mask_weights.to_tensor(shape=target_shape)
```

Reply:
done!